sideload/sideload.go

```diff
@@ -0,0 +1,416 @@
+// constraint.
+type SideLoader[T header] struct {
+	// LoaderInternal contains basic loader functionalities such as
+	// writing and validating headers, along with configurations for
+	// sideloading and validation processes.
+	LoaderInternal[T]
+
+	// source represents the data source from which headers are sideloaded.
+	// It is specific to the sideloading process, encapsulating the logic
+	// for fetching and decoding sideloaded headers.
+	source LoaderSource[T]
+
+	// chkptTracker keeps track of the current checkpoint index during the
+	// sideloading process.
+	chkptTracker int32
+
+	// haltHeight specifies a blockchain height at which the sideloading
+	// process should be halted.
+	haltHeight uint32
+
+	// haltHeightFunc is a function that dynamically determines the halt height
+	// for the sideloading process.
+	haltHeightFunc func() uint32
+}
+
+// NewBlockHeaderLoader initializes a block header Loader based on the source
+// type of the reader config.
+func NewBlockHeaderLoader(
+	cfg *LoaderConfig[*wire.BlockHeader]) (*SideLoader[*wire.BlockHeader],
+	error) {
+
+	var (
+		source LoaderSource[*wire.BlockHeader]
+		err    error
+	)
+
+	switch {
+	case cfg.SourceType == Binary:
+		source, err = newBinaryBlkHdrLoader(cfg.Reader)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, ErrUnsupportedSourceType
+	}
+
+	headerChain := source.HeadersChain()
+
+	if headerChain != cfg.Chain {
+		return nil, ErrBitcoinNetworkMismatchFmt(
+			cfg.Chain, headerChain,
+		)
+	}
+
+	return &SideLoader[*wire.BlockHeader]{
+		LoaderInternal: cfg.LoaderInternal,
+		source:         source,
+		haltHeightFunc: func() uint32 {
+			return source.EndHeight()
+		},
+	}, nil
+}
+
+// obtainRangeFetchFunc returns the next number of headers to be fetched and
+// the next checkpoint header and checkpoint height, if skipVerify is false.
+func (s *SideLoader[T]) obtainRangeFetchFunc() func(
+	curHeight uint32) (uint32, *chainhash.Hash) {
+
+	// If we are to verify headers and the length of checkpoints is
+	// greater than zero, we want to fetch headers from our current height to
+	// our next checkpoint.
+	if !s.SkipVerify && s.Chkpt.Len() > 0 {
+		return func(curHeight uint32) (uint32, *chainhash.Hash) {
+			chkpt := s.Chkpt
+
+			// Check if idx is out of range
+			if s.chkptTracker >= chkpt.Len() {
+				return 0, nil
+			}
+			checkpointHeight, checkpointHash := chkpt.
+				FetchCheckpoint(s.chkptTracker)
+
+			if checkpointHeight > s.source.EndHeight() {
+				return curHeight, nil
+			}
+
+			s.chkptTracker++
+
+			return checkpointHeight, checkpointHash
+		}
+	}
+
+	// If we do not want to Verify headers we use whatever
+	//  `sideloadRange` was given to us by the caller.
+	return func(curHeight uint32) (uint32, *chainhash.Hash) {
+		if curHeight >= s.haltHeight {
+			return curHeight, nil
+		}
+		if curHeight+s.SideloadRange > s.haltHeight {
+			return s.haltHeight - curHeight, nil
+		}
+
+		return curHeight + s.SideloadRange, nil
+	}
+}
+
```

Comment:
I am thinking there should be nothing like sideload range and we just fetch from checkpoint to checkpoint just as the blockmanager currently does whether we are verifying or not.